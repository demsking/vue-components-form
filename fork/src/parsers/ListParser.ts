import { AbstractParser } from '@/parsers/AbstractParser';
import { ListField, ListItem, ScalarDescriptor } from '@/types';

export class ListParser extends AbstractParser<any, ScalarDescriptor, ListField> {
  readonly enums: any[] = [];

  get defaultComponent() {
    return this.descriptor.kind
      ? this.options.descriptorConstructor<ScalarDescriptor>(this.schema, this.descriptor.kind).component
      : this.options.descriptorConstructor(this.schema, 'list').component;
  }

  get items(): ListItem[] {
    return this.enums.map((item, i) => ({
      index: i,
      value: item,
      label: this.descriptor.labels
        ? this.descriptor.labels[item] || item
        : item
    }));
  }

  parse() {
    if (this.schema.enum instanceof Array) {
      this.enums.push(...this.schema.enum);
    }

    this.field.kind = 'list';
    this.field.items = this.items;

    this.parseField();
  }

  parseValue(data: any): string {
    return typeof data !== 'undefined' ? data : '';
  }
}
