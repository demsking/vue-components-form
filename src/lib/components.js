import { INPUT_TYPES } from './parser'

const BLOCK_TYPES = [INPUT_TYPES.TEXTAREA, INPUT_TYPES.SELECT]

const Input = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const element = BLOCK_TYPES.includes(field.attrs.type)
      ? h(field.attrs.type, data, slots().default)
      : h('input', data)
    const nodes = [ element ]

    if (field.description) {
      nodes.push(h('small', field.description))
    }

    if (!field.label || (field.isArrayField && !BLOCK_TYPES.includes(field.attrs.type))) {
      if (nodes.length === 1) {
        return nodes[0]
      }
      return nodes
    }

    return h(Label, data, nodes)
  }
}

const CheckboxGroup = {
  functional: true,
  render (h, { data, slots }) {
    const name = data.field.attrs.name
    const description = data.field.description

    const children = [ ...slots().default ]

    if (description) {
      children.unshift(h('legend', description))

      delete data.field.description
    }

    return h('fieldset', { attrs: { name } }, children)
  }
}

const ArrayInputs = {
  functional: true,
  render (h, { slots }) {
    return h('div', {
      attrs: {
        'data-fs-array-inputs': true
      }
    }, slots().default)
  }
}

const Label = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const attrs = {
      'data-fs-field': field.attrs.id,
      'data-fs-required': field.attrs.required
    }

    return h('div', { attrs }, [
      h('label', {
        attrs: {
          for: field.attrs.id
        }
      }, field.label),
      h('div', {
        attrs: {
          'data-fs-field-input': field.attrs.id
        }
      }, slots().default)
    ])
  }
}

const Fieldset = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const newItemButton = data.newItemButton
    const nodes = []

    if (field.isArrayField) {
      if (newItemButton) {
        nodes.push(h(ArrayInputs, slots().default))
        nodes.push(h(ArrayButton, newItemButton))
      } else {
        nodes.push(h(CheckboxGroup, data, slots().default))
      }
    } else {
      nodes.push(h('fieldset', slots().default))
    }

    if (field.description) {
      nodes.push(h('small', field.description))
    }

    if (!field.label) {
      if (nodes.length === 1) {
        return nodes[0]
      }
      return nodes
    }

    return h(Label, data, nodes)
  }
}

const ArrayButton = {
  functional: true,
  render (h, { data }) {
    return h('button', {
      attrs: { type: 'button', ...data.props },
      on: data.on
    }, 'Add')
  }
}

const ErrorElement = {
  functional: true,
  render (h, { slots }) {
    return h('div', {
      attrs: {
        'data-fs-error': true
      }
    }, slots().default)
  }
}

const TAGS = {
  title: 'h1',
  description: 'p',
  formwrapper: 'div',
  form: 'form',
  text: Input,
  option: 'option',
  fieldset: Fieldset,
  error: ErrorElement
}

export class Components {
  constructor () {
    this.$ = {}

    for (let type in TAGS) {
      this.set(type, TAGS[type], true)
    }
  }

  set (type, component, native = false) {
    this.$[type] = { type, native, component }
  }

  input ({ field, fieldParent = null, listeners = {} }) {
    const { type } = field.attrs
    const element = field.hasOwnProperty('items') && groupedArrayTypes.includes(type)
      ? this.$.fieldset
      : this.$[type] || this.$[INPUT_TYPES.TEXT]

    const data = {
      field,
      fieldParent,
      components: this.components,
      attrs: {},
      props: {},
      domProps: {},
      [argName(element)]: { ...field.attrs },
      on: listeners
    }

    return { element, data }
  }
}

export function argName (el) {
  return el.native ? 'attrs' : 'props'
}

export const groupedArrayTypes = [
  INPUT_TYPES.RADIO, INPUT_TYPES.CHECKBOX, 'input', INPUT_TYPES.TEXTAREA
]

export const fieldTypesAsNotArray = [
  INPUT_TYPES.RADIO, INPUT_TYPES.TEXTAREA, INPUT_TYPES.SELECT
]

export const inputName = (field, index) => `${field.attrs.name}-${index}`
