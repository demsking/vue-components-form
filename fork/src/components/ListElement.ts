import { CreateInput } from '@/lib/CreateInput';
import { HelperElement } from "@/components/HelperElement";
import { FieldElement } from "@/components/FieldElement";
import { ListComponent } from '@/types';

export const ListElement: ListComponent = {
  name: 'ListElement',
  functional: true,
  render(h, { data, props }) {
    const children = props.field.items.map(({ label, value, selected }) => {
      const attrs = { value, selected };

      return h('option', { attrs }, label);
    });

    return h(FieldElement, data, [
      CreateInput(h, 'select', data, children),
      h(HelperElement, data)
    ]);
  }
};
