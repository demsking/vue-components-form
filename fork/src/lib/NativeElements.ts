import { Components } from '@/lib/Components';
import { InputElement } from '@/components/InputElement';
import { FieldsetElement } from '@/components/FieldsetElement';
import { SelectElement } from '@/components/SelectElement';
import { TextareaElement } from '@/components/TextareaElement';

export const NativeElements = new Components();

NativeElements.set('boolean', InputElement);
NativeElements.set('string', InputElement);
NativeElements.set('radio', InputElement);
NativeElements.set('enum', FieldsetElement);
NativeElements.set('number', InputElement);
NativeElements.set('integer', InputElement);
NativeElements.set('object', FieldsetElement);
NativeElements.set('list', SelectElement);
NativeElements.set('textarea', TextareaElement);
