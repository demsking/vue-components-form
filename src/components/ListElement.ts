import { CreateInput } from '@/lib/CreateInput';
import { FieldElement } from '@/components/FieldElement';
import { ListComponent } from '@/types';

export const ListElement: ListComponent = {
  name: 'ListElement',
  functional: true,
  render(h, { data, props }) {
    const items = [ ...props.field.items ];

    items.unshift({
      label: props.field.input.props.placeholder,
      value: '',
      selected: false
    });

    const children = items.map(({ label, value, selected }) => {
      return h('option', { attrs: { value, selected } }, label);
    });

    return h(FieldElement, data, [
      CreateInput(h, 'select', data, children, 'change')
    ]);
  }
};
