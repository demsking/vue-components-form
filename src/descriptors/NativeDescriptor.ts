import { Schema } from '@/lib/Schema';
import { JsonSchema } from '@/types/jsonschema';
import { NativeElements } from '@/lib/NativeElements';

import {
  FieldKind,
  ScalarDescriptor,
  ObjectDescriptor,
  ArrayDescriptor,
  DescriptorInstance
} from '@/types';

export const NativeDescriptor = {
  get<T = DescriptorInstance>(schema: JsonSchema, kind?: FieldKind): T {
    const element = NativeElements.get(kind || schema.type);

    if (Schema.isScalar(schema)) {
      const descriptor: ScalarDescriptor = {
        label: schema.title,
        description: schema.description,
        component: element,
        attrs: {},
        props: {},
        labels: {}
      };

      return descriptor as T;
    }

    if (schema.type === 'array') {
      const descriptor: ArrayDescriptor = {
        label: schema.title,
        description: schema.description,
        component: element,
        attrs: {},
        props: {},
        items: [],
        addButtonLabel: '+'
      };

      return descriptor as unknown as T;
    }

    const descriptor: ObjectDescriptor = {
      label: schema.title,
      description: schema.description,
      component: element,
      attrs: {},
      props: {},
      order: []
    };

    return descriptor as T;
  }
};
