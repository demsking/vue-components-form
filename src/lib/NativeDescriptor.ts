import { Schema } from '@/lib/Schema';
import { JsonSchema } from '@/types/jsonschema';
import { NativeElements } from '@/lib/NativeElements';

import {
  FieldKind,
  ScalarDescriptor,
  ObjectDescriptor,
  ArrayDescriptor,
  DescriptorInstance
} from '@/types';

export const NativeDescriptor = {
  kind(schema: JsonSchema) {
    return schema.enum
      ? schema.enum.length > 4 ? 'list' : 'enum'
      : schema.type;
  },
  get<T = DescriptorInstance>(schema: JsonSchema, kind?: FieldKind): T {
    const kindUsed = typeof kind === 'undefined'
      ? NativeDescriptor.kind(schema)
      : kind;

    const element = NativeElements.get(kindUsed);

    if (Schema.isScalar(schema)) {
      const descriptor: ScalarDescriptor = {
        kind: kindUsed,
        label: schema.title,
        helper: schema.description,
        component: element,
        attrs: {},
        props: {},
        items: {}
      };

      return descriptor as T;
    }

    if (schema.type === 'array') {
      const descriptor: ArrayDescriptor = {
        kind: kindUsed,
        label: schema.title,
        helper: schema.description,
        component: element,
        attrs: {},
        props: {},
        items: [],
        addButtonLabel: '+'
      };

      return descriptor as unknown as T;
    }

    const descriptor: ObjectDescriptor = {
      kind: kindUsed,
      label: schema.title,
      helper: schema.description,
      component: element,
      attrs: {},
      props: {},
      order: []
    };

    return descriptor as T;
  }
};
