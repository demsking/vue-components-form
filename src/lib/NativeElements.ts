import { Components } from '@/lib/Components';
import { InputElement } from '@/components/InputElement';
import { CheckboxElement } from '@/components/CheckboxElement';
import { ArrayElement } from '@/components/ArrayElement';
import { FieldsetElement } from '@/components/FieldsetElement';
import { ListElement } from '@/components/ListElement';
import { TextareaElement } from '@/components/TextareaElement';
import { BooleanElement } from '@/components/BooleanElement';

export const NativeElements = new Components();

NativeElements.set('array', ArrayElement);
NativeElements.set('boolean', BooleanElement);
NativeElements.set('string', InputElement);
NativeElements.set('radio', InputElement);
NativeElements.set('checkbox', CheckboxElement);
NativeElements.set('enum', FieldsetElement);
NativeElements.set('number', InputElement);
NativeElements.set('integer', InputElement);
NativeElements.set('object', FieldsetElement);
NativeElements.set('list', ListElement);
NativeElements.set('textarea', TextareaElement);
