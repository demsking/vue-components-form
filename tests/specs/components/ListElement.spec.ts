import { mount } from '@vue/test-utils';
import { ListElement } from '@/components/ListElement';
import { ListParser } from '@/parsers/ListParser';
import { NativeDescriptor } from '@/lib/NativeDescriptor';
import { NativeElements } from '@/lib/NativeElements';

const options: any = {
  schema: {
    type: 'string',
    title: 'Character',
    description: 'Your character',
    enum: ['goku', 'freezer']
  },
  model: 'freezer',
  id: 'id-character',
  name: 'character',
  onChange: jest.fn(),
  descriptor: {
    items: {
      goku: {
        label: 'Goku'
      },
      freezer: {
        label: 'Freezer'
      }
    }
  },
  descriptorConstructor: new NativeDescriptor(NativeElements)
};

const parser = new ListParser(options);

parser.parse();

const context: any = {
  attrs: parser.field.input.attrs,
  props: {
    field: parser.field
  }
};

describe('components/ListElement', () => {
  it('should successfully render component', () => {
    const wrapper = mount(ListElement, { context });
    const expected = '<div data-fs-kind="list" data-fs-field="character"><label id="id-character-label" for="id-character">Character</label><div data-fs-wrapper="2"><div data-fs-input="list"><select id="id-character" name="character" aria-labelledby="id-character-label" aria-describedby="id-character-helper"><option value=""></option><option value="goku">Goku</option><option value="freezer" selected="selected">Freezer</option></select></div><p id="id-character-helper">Your character</p></div></div>';

    expect(wrapper.html()).toBe(expected);
  });

  it('should successfully emit input event', () => {
    const wrapper = mount(ListElement, { context });
    const select = wrapper.find('select');
    const option: any = wrapper.find('option[value=goku]');

    option.element.selected = true;

    select.trigger('change');

    const [ [ initialValue ], [ changedValue ] ] = options.onChange.mock.calls;

    expect(initialValue).toEqual('freezer');
    expect(changedValue).toEqual('goku');
  });
});
