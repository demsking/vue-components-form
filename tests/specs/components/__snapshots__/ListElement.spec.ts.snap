// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ListElement should successfully render component 1`] = `
<div data-fs-kind="list" data-fs-type="list" data-fs-field="character"><label id="id-character-label" for="id-character">Character</label>
  <div data-fs-wrapper="2">
    <div data-fs-input="list"><select id="id-character" name="character" aria-labelledby="id-character-label" aria-describedby="id-character-helper">
        <option value=""></option>
        <option value="goku">Goku</option>
        <option value="freezer" selected="selected">Freezer</option>
      </select></div>
    <p id="id-character-helper" data-fs-helper="true">Your character</p>
  </div>
</div>
`;
